src/batch/src/executor/s3_file_scan.rs

```diff
@@ -0,0 +1,163 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ops::Range;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use bytes::Bytes;
+use futures_async_stream::try_stream;
+use futures_util::future::BoxFuture;
+use futures_util::stream::StreamExt;
+use futures_util::TryFutureExt;
+use hashbrown::HashMap;
+use iceberg::io::{
+    FileIOBuilder, FileMetadata, FileRead, S3_ACCESS_KEY_ID, S3_REGION, S3_SECRET_ACCESS_KEY,
+};
+use parquet::arrow::async_reader::{AsyncFileReader, MetadataLoader};
+use parquet::arrow::{ParquetRecordBatchStreamBuilder, ProjectionMask};
+use parquet::file::metadata::ParquetMetaData;
+use risingwave_common::array::arrow::IcebergArrowConvert;
+use risingwave_common::catalog::Schema;
+
+use crate::error::BatchError;
+use crate::executor::{DataChunk, Executor};
+
+#[derive(PartialEq, Debug)]
+pub enum FileFormat {
+    Parquet,
+}
+
+/// S3 file scan executor. Currently only support parquet file format.
+pub struct S3FileScanExecutor {
+    file_format: FileFormat,
+    location: String,
+    s3_region: String,
+    s3_access_key: String,
+    s3_secret_key: String,
+    batch_size: usize,
+    schema: Schema,
+    identity: String,
+}
+
+impl Executor for S3FileScanExecutor {
+    fn schema(&self) -> &risingwave_common::catalog::Schema {
+        &self.schema
+    }
+
+    fn identity(&self) -> &str {
+        &self.identity
+    }
+
+    fn execute(self: Box<Self>) -> super::BoxedDataChunkStream {
+        self.do_execute().boxed()
+    }
+}
+
+impl S3FileScanExecutor {
+    #![expect(dead_code)]
+    pub fn new(
+        file_format: FileFormat,
+        location: String,
+        s3_region: String,
+        s3_access_key: String,
+        s3_secret_key: String,
+        batch_size: usize,
+        schema: Schema,
+        identity: String,
+    ) -> Self {
+        Self {
+            file_format,
+            location,
+            s3_region,
+            s3_access_key,
+            s3_secret_key,
+            batch_size,
+            schema,
+            identity,
+        }
+    }
+
+    #[try_stream(ok = DataChunk, error = BatchError)]
+    async fn do_execute(self: Box<Self>) {
+        assert_eq!(self.file_format, FileFormat::Parquet);
+
+        let mut props = HashMap::new();
+        props.insert(S3_REGION, self.s3_region.clone());
+        props.insert(S3_ACCESS_KEY_ID, self.s3_access_key.clone());
+        props.insert(S3_SECRET_ACCESS_KEY, self.s3_secret_key.clone());
+
+        let file_io_builder = FileIOBuilder::new("s3");
+        let file_io = file_io_builder.with_props(props.into_iter()).build()?;
+        let parquet_file = file_io.new_input(&self.location)?;
```

Comment:
Is there any reason of using `iceberg::io` to access the object storage, instead of OpenDAL or AWS client?

Reply:
Initially, I implemented it with OpenDAL, but found that to read a parquet file, we need to reimplement a bunch of logic that has already been implemented in iceberg-rust, so I don't want to reinvent the wheel again. BTW, iceberg `FileIO` looks good to me, because in iceberg scan, we need to use this interface to avoid resolving iceberg metadata again in the compute node while scanning parquet files.

Reply:
I see. I noticed that `iceberg::io` actually use OpenDAL under the hood. 

https://github.com/apache/iceberg-rust/blob/48f9e3e8bff0224569da8dd5beb89ed0c8cc4513/crates/iceberg/src/io/storage.rs#L76

However, don't know why the implementation limits the storage to be AWS S3. Do you know?

Reply:
I think the core module of iceberg-rust is still under development, so only AWS S3 is supported currently.